marlowe-dashboard-client/src/Play/State.purs

```diff
@@ -0,0 +1,103 @@
+module Play.State
+  ( mkInitialState
+  , handleAction
+  ) where
+
+import Prelude
+import Contract.State (defaultState, handleAction) as Contract
+import Contract.Types (_executionState)
+import Control.Monad.State.Class (get)
+import Data.Foldable (for_)
+import Data.Lens (assign, modifying, set, view)
+import Data.Lens.Extra (peruse)
+import Data.Maybe (Maybe(..))
+import Effect.Aff.Class (class MonadAff)
+import Halogen (HalogenM, modify_)
+import Halogen.Extra (mapMaybeSubmodule)
+import MainFrame.Lenses (_card, _playState, _screen)
+import MainFrame.Types (Action(..), State) as MainFrame
+import MainFrame.Types (ChildSlots, Msg)
+import Marlowe.Execution (_contract)
+import Marlowe.Extended (fillTemplate, toCore)
+import Marlowe.Semantics (PubKey)
+import Play.Lenses (_contractState, _menuOpen, _templateState)
+import Play.Types (Action(..), Card(..), ContractStatus(..), Screen(..), State)
+import Template.Lenses (_extendedContract, _template, _templateContent)
+import Template.State (defaultState, handleAction, mkInitialState) as Template
+import Template.Types (Action(..), Screen(..)) as Template
+
+mkInitialState :: PubKey -> State
+mkInitialState pubKeyHash =
+  { wallet: pubKeyHash
+  , menuOpen: false
+  , screen: ContractsScreen Running
+  , card: Nothing
+  , templateState: Template.defaultState
+  , contractState: Contract.defaultState
+  }
+
+-- Some actions are handled in `MainFrame.State` because they involve
+-- modifications of that state. See Note [State].
+handleAction :: forall m. MonadAff m => Action -> HalogenM MainFrame.State MainFrame.Action ChildSlots Msg m Unit
+handleAction (SetNewWalletNickname _) = pure unit -- handled in `MainFrame.State`
+
+handleAction (SetNewWalletKey key) = pure unit -- handled in `MainFrame.State`
+
+handleAction AddNewWallet = pure unit -- handled in `MainFrame.State`
+
+handleAction PutdownWallet = pure unit -- handled in `MainFrame.State`
+
+handleAction ToggleMenu = modifying (_playState <<< _menuOpen) not
+
+handleAction (SetScreen screen) =
+  modify_
+    $ (_playState <<< set _menuOpen) false
+    <<< (_playState <<< set _card) Nothing
+    <<< (_playState <<< set _screen) screen
+
+handleAction (SetCard card) = do
+  previousCard <- peruse (_playState <<< _card)
+  assign (_playState <<< _card) card
+  for_ previousCard $ const $ assign (_playState <<< _menuOpen) false
+
+handleAction (ToggleCard card) = do
+  mCurrentCard <- peruse (_playState <<< _card)
+  case mCurrentCard of
+    Just currentCard
+      | currentCard == Just card -> handleAction $ SetCard Nothing
```

Comment:
since you are already using `peruse` I would say it's a bit clearer to have
```purescript
  mCurrentCard <- peruse (_playState <<< _card <<<< _Just)
  case mCurrentCard of
    Just currentCard
      | currentCard == card -> handleAction $ SetCard Nothing
```

Reply:
👍